#### Setup

*Connecting to the PostgreSQL database and loading the `ipython-sql` extension so SQL queries can run directly inside the notebook (via the `%%sql` magic).*

In [1]:
import pandas as pd
from sqlalchemy import create_engine
%load_ext sql

In [2]:
engine=create_engine('postgresql+psycopg2://postgres:password@localhost:5432/medika_sejahtera')
connection=engine.connect()
print("Connected:", connection.closed == False)

Connected: True


In [3]:
%sql postgresql+psycopg2://postgres:password@localhost:5432/medika_sejahtera

'Connected: postgres@medika_sejahtera'

#### Business Question

1. What reason code most frequently causes product returns?
2. Which products and product categories are returned most often (by frequency)?
3. What is the total financial loss from returns across 2024, and which products/categories account for the largest share?
4. Which customer type (healthcare institution type) has the most returns?
5. Which region records the highest number of return cases?
6. What does the return trend look like across 2024 — is there a monthly or seasonal pattern?

# 1. Product & Reason Analysis

*Baseline: total count and share (%) of returns by reason code, across the full dataset.*

In [4]:
%%sql

SELECT
    return_reason_cleaned,
    COUNT (*) AS total_record
FROM
    return_cleaned
GROUP BY
    return_reason_cleaned
ORDER BY
    total_record DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
4 rows affected.


return_reason_cleaned,total_record
Salah Input Sistem,28
Salah Pesan Customer,20
Barang Rusak,7
Kadaluarsa,4


In [27]:
%%sql
WITH total AS
(
    SELECT
        COUNT(*) AS total_records
    FROM
        return_cleaned
)

SELECT
    rc.return_reason_cleaned,
    COUNT(*) AS total_record,
    tl.total_records AS total_data,
    ROUND((COUNT(*)::numeric / NULLIF(tl.total_records,0)::numeric) * 100,2) AS rate
FROM
    return_cleaned rc
CROSS JOIN
    total tl
GROUP BY
    rc.return_reason_cleaned,
    total_data
ORDER BY
    total_record DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
4 rows affected.


return_reason_cleaned,total_record,total_data,rate
Salah Input Sistem,28,59,47.46
Salah Pesan Customer,20,59,33.90
Barang Rusak,7,59,11.86
Kadaluarsa,4,59,6.78


**Screening**

System Input Error (Salah Input Sistem) (SIS, 47.46%) and Customer Ordering Error (Salah Pesan Customer) (SPC, 33.90%) are both major contributors, together accounting for 81.36% of all cases. But they come from two different failure points — one from an internal process, the other from the customer's side when placing an order — so they need two separate fixes, not a single blanket solution.

#### Screening 1D

(baseline per reason code, then broken down by region / quarter / category / customer type / product)

##### *Reason code breakdown by quarter.*

In [35]:
%%sql

WITH reason_per_quarter AS (
    SELECT
        return_reason_cleaned,
        CASE
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 1 AND 3 THEN 'Q1'
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 4 AND 6 THEN 'Q2'
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 7 AND 9 THEN 'Q3'
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 10 AND 12 THEN 'Q4'
        END AS quarter,
        COUNT(*) AS total_record_filter
    FROM 
        return_cleaned
    GROUP BY 
        return_reason_cleaned, 
        quarter
)
SELECT
    return_reason_cleaned,
    quarter,
    total_record_filter,
    SUM(total_record_filter) OVER (PARTITION BY quarter) AS total_per_quarter,
    ROUND(
        total_record_filter * 100.0 / SUM(total_record_filter) OVER (PARTITION BY quarter),
        2
    ) AS row_pct
FROM 
    reason_per_quarter
ORDER BY 
    return_reason_cleaned,
    quarter, 
    row_pct DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
14 rows affected.


return_reason_cleaned,quarter,total_record_filter,total_per_quarter,row_pct
Barang Rusak,Q1,3,8,37.50
Barang Rusak,Q2,2,15,13.33
Barang Rusak,Q3,2,14,14.29
Kadaluarsa,Q1,1,8,12.50
Kadaluarsa,Q3,2,14,14.29
Kadaluarsa,Q4,1,22,4.55
Salah Input Sistem,Q1,2,8,25.00
Salah Input Sistem,Q2,10,15,66.67
Salah Input Sistem,Q3,6,14,42.86
Salah Input Sistem,Q4,10,22,45.45


In [36]:
%%sql

SELECT
    CASE
        WHEN EXTRACT(MONTH FROM invoice_date) BETWEEN 1 AND 3 THEN 'Q1'
        WHEN EXTRACT(MONTH FROM invoice_date) BETWEEN 4 AND 6 THEN 'Q2'
        WHEN EXTRACT(MONTH FROM invoice_date) BETWEEN 7 AND 9 THEN 'Q3'
        WHEN EXTRACT(MONTH FROM invoice_date) BETWEEN 10 AND 12 THEN 'Q4'
    END AS quarter,
    COUNT(*) AS total_data
FROM 
    invoice_data
GROUP BY 
    quarter
ORDER BY 
    quarter

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
4 rows affected.


quarter,total_data
Q1,149
Q2,171
Q3,179
Q4,204


*Return rate relative to invoice volume per quarter, focused on SIS and SPC, with quarter-over-quarter growth — this is the query behind the screening findings below.*

In [37]:
%%sql
WITH return_by_reason_quarter AS (
    SELECT
        return_reason_cleaned,
        CASE
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 1 AND 3 THEN 'Q1'
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 4 AND 6 THEN 'Q2'
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 7 AND 9 THEN 'Q3'
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 10 AND 12 THEN 'Q4'
        END AS quarter,
        COUNT(*) AS return_count
    FROM return_cleaned
    GROUP BY return_reason_cleaned, quarter
),
invoice_by_quarter AS (
    SELECT
        CASE
            WHEN EXTRACT(MONTH FROM invoice_date) BETWEEN 1 AND 3 THEN 'Q1'
            WHEN EXTRACT(MONTH FROM invoice_date) BETWEEN 4 AND 6 THEN 'Q2'
            WHEN EXTRACT(MONTH FROM invoice_date) BETWEEN 7 AND 9 THEN 'Q3'
            WHEN EXTRACT(MONTH FROM invoice_date) BETWEEN 10 AND 12 THEN 'Q4'
        END AS quarter,
        COUNT(*) AS invoice_count
    FROM invoice_data
    GROUP BY quarter
),
rate_by_quarter AS (
    SELECT
        rq.return_reason_cleaned,
        rq.quarter,
        rq.return_count,
        iq.invoice_count,
        ROUND(
            (rq.return_count::numeric * 100) / NULLIF(iq.invoice_count, 0)::numeric,
            2
        ) AS rate_pct
    FROM return_by_reason_quarter rq
    JOIN invoice_by_quarter iq ON rq.quarter = iq.quarter
)
SELECT
    return_reason_cleaned,
    quarter,
    return_count,
    invoice_count,
    rate_pct,
    ROUND(
        (rate_pct - LAG(rate_pct) OVER (PARTITION BY return_reason_cleaned ORDER BY quarter))
        / NULLIF(LAG(rate_pct) OVER (PARTITION BY return_reason_cleaned ORDER BY quarter), 0)
        * 100,
        2
    ) AS growth_pct_qoq
FROM rate_by_quarter
WHERE return_reason_cleaned IN ('Salah Input Sistem', 'Salah Pesan Customer')
ORDER BY return_reason_cleaned, quarter;

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
8 rows affected.


return_reason_cleaned,quarter,return_count,invoice_count,rate_pct,growth_pct_qoq
Salah Input Sistem,Q1,2,149,1.34,None
Salah Input Sistem,Q2,10,171,5.85,336.57
Salah Input Sistem,Q3,6,179,3.35,-42.74
Salah Input Sistem,Q4,10,204,4.90,46.27
Salah Pesan Customer,Q1,2,149,1.34,None
Salah Pesan Customer,Q2,3,171,1.75,30.60
Salah Pesan Customer,Q3,4,179,2.23,27.43
Salah Pesan Customer,Q4,11,204,5.39,141.70


**Screening**  


Q2 (n=15) — SIS spikes (66.67%, +19.21 pts), SPC dips (20%, -13.90 pts) → worth digging into

Q4 (n=22) — SPC spikes (50%, +16.10 pts), SIS roughly at baseline (45.45%, -2.01 pts, negligible) → worth digging into

Q3 (n=14) — all reasons close to baseline → clean screening

Q1 (n=8, individual cells n=1-3) — large deviations on paper (SIS -22.46, Damaged Goods +25.64), but cell sizes are too small → minor observation 

##### *Reason code breakdown by region.*

In [63]:
%%sql

WITH reason_count AS (
    SELECT 
        region_cleaned,
        return_reason_cleaned,
        COUNT(*) AS n
    FROM 
        return_cleaned
    GROUP BY  
        region_cleaned, 
        return_reason_cleaned
)
SELECT 
    region_cleaned,
    return_reason_cleaned,
    n,
    SUM(n) OVER (PARTITION BY region_cleaned) AS total_per_region,
    ROUND(
        n * 100.0 / SUM(n) OVER (PARTITION BY region_cleaned), 
        2
    ) AS row_pct
FROM 
    reason_count
ORDER BY
    region_cleaned,
    row_pct DESC


 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
25 rows affected.


region_cleaned,return_reason_cleaned,n,total_per_region,row_pct
Bekasi,Salah Input Sistem,5,10,50.00
Bekasi,Salah Pesan Customer,4,10,40.00
Bekasi,Kadaluarsa,1,10,10.00
Bogor,Salah Input Sistem,2,4,50.00
Bogor,Salah Pesan Customer,2,4,50.00
Depok,Kadaluarsa,1,3,33.33
Depok,Barang Rusak,1,3,33.33
Depok,Salah Input Sistem,1,3,33.33
Jakarta Barat,Salah Input Sistem,8,13,61.54
Jakarta Barat,Salah Pesan Customer,3,13,23.08


**Screening**  
West Jakarta (n=13) — SIS stands out (61.54%, +14.08 pts), SPC low (23.08%, -10.82 pts) → worth digging into

South Jakarta (n=10) — both SIS (50%, +2.54 pts) and SPC (50%, +16.10 pts) above baseline → worth digging into

East Jakarta (n=6) — SPC stands out (50%, +16.10 pts) → worth digging into (small n)

Tangerang (n=5, Damaged Goods cell n=3) — Damaged Goods high (60%, baseline 11.86%, +48.14 pts) → worth digging into (very small n, treat with caution)

Bekasi (n=10) — all reasons close to baseline → clean screening

Central Jakarta, Bogor, Depok, North Jakarta (n=3-5 per region, individual cells n=1-2) → n too small to draw conclusions, minor observation only

##### *Reason code breakdown by category.*

In [65]:
%%sql

WITH category_count AS (
    SELECT
        kategori,
        return_reason_cleaned,
        COUNT (*) AS total_records
    FROM
        return_cleaned
    GROUP BY
        kategori,
        return_reason_cleaned
)


SELECT
    kategori,
    return_reason_cleaned,
    total_records,
    SUM(total_records) OVER (PARTITION BY kategori) AS total_per_category,
    ROUND (total_records * 100.0 / SUM(total_records) OVER (PARTITION BY kategori),2) AS row_pct
FROM
    category_count
ORDER BY
    kategori,
    total_records DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
17 rows affected.


kategori,return_reason_cleaned,total_records,total_per_category,row_pct
Bedah & Tindakan,Salah Input Sistem,5,11,45.45
Bedah & Tindakan,Salah Pesan Customer,5,11,45.45
Bedah & Tindakan,Kadaluarsa,1,11,9.09
Consumable,Salah Input Sistem,6,13,46.15
Consumable,Salah Pesan Customer,4,13,30.77
Consumable,Kadaluarsa,2,13,15.38
Consumable,Barang Rusak,1,13,7.69
Diagnostik,Salah Input Sistem,9,16,56.25
Diagnostik,Salah Pesan Customer,4,16,25.00
Diagnostik,Barang Rusak,2,16,12.50


**Screening**  
Diagnostics — SIS stands out (56.25%, baseline 47.46%, +8.79 pts), SPC low (25%, baseline 33.90%, -8.90 pts) → worth digging into

Laboratory — SIS stands out (62.5%, baseline 47.46%, +15.04 pts), SPC low (25%, baseline 33.90%, -8.90 pts) → worth digging into (n=8, small — read alongside Diagnostics as a related category, not on its own)

Surgery & Procedures (Bedah & Tindakan) — SPC stands out (45.45%, baseline 33.90%, +11.55 pts), SIS close to baseline (45.45% vs. 47.46%, -2.01 pts, negligible) → worth digging into

Nursing & Inpatient Care (Perawatan & Rawat Inap) — SPC stands out (45.45%, baseline 33.90%, +11.55 pts), SIS low (27.27%, baseline 47.46%, -20.19 pts) → worth digging into

##### *Reason code breakdown by customer type.*

In [66]:
%%sql

WITH customer_type_count AS (
    SELECT
            customer_type_cleaned,
            return_reason_cleaned,
            COUNT (*) AS total_records
        FROM
            return_cleaned
        GROUP BY
            customer_type_cleaned,
            return_reason_cleaned

)

SELECT
    customer_type_cleaned,
    return_reason_cleaned,
    total_records,
    SUM(total_records) OVER (PARTITION BY customer_type_cleaned) AS total_per_customer_type,
    ROUND (total_records * 100.0 / SUM(total_records) OVER (PARTITION BY customer_type_cleaned),2) AS row_pct
FROM
    customer_type_count
ORDER BY
    customer_type_cleaned,
    total_records DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
12 rows affected.


customer_type_cleaned,return_reason_cleaned,total_records,total_per_customer_type,row_pct
Klinik,Salah Input Sistem,9,19,47.37
Klinik,Salah Pesan Customer,8,19,42.11
Klinik,Barang Rusak,1,19,5.26
Klinik,Kadaluarsa,1,19,5.26
RS Pemerintah,Salah Input Sistem,8,13,61.54
RS Pemerintah,Barang Rusak,2,13,15.38
RS Pemerintah,Salah Pesan Customer,2,13,15.38
RS Pemerintah,Kadaluarsa,1,13,7.69
RS Swasta,Salah Input Sistem,11,27,40.74
RS Swasta,Salah Pesan Customer,10,27,37.04


**Screening**  
Government Hospital (RS Pemerintah) — SIS stands out (61.54%, baseline 47.46%, +14.08 pts), SPC low (15.38%, baseline 33.90%) → worth digging into

Clinic — SPC deviates (42.11%, baseline 33.90%, +8.21 pts), SIS at baseline (47.37% vs. 47.46%, negligible) → worth digging into (SPC specifically)

Private Hospital (RS Swasta) — SIS (40.74%, -6.72 pts) and SPC (37.04%, +3.14 pts), both close to baseline → clean screening, no meaningful deviation

##### *Reason code breakdown at the individual product (SKU) level.*

In [67]:
%%sql

WITH product_count AS (
    SELECT
            sku_name,
            return_reason_cleaned,
            COUNT (*) AS total_records
        FROM
            return_cleaned
        GROUP BY
            sku_name,
            return_reason_cleaned

)

SELECT
    sku_name,
    return_reason_cleaned,
    total_records,
    SUM(total_records) OVER (PARTITION BY sku_name) AS total_per_product,
    ROUND (total_records * 100.0 / SUM(total_records) OVER (PARTITION BY sku_name),2) AS row_pct
FROM
    product_count
ORDER BY
    sku_name,
    total_records DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
50 rows affected.


sku_name,return_reason_cleaned,total_records,total_per_product,row_pct
Baju Pasien Dewasa,Barang Rusak,1,2,50.00
Baju Pasien Dewasa,Salah Input Sistem,1,2,50.00
Bantal Angin Anti Decubitus,Salah Pesan Customer,1,1,100.00
Bedpan Plastik Dewasa,Salah Input Sistem,1,1,100.00
Benang Catgut 2-0,Salah Input Sistem,2,3,66.67
Benang Catgut 2-0,Salah Pesan Customer,1,3,33.33
Duk Operasi 80x90cm,Kadaluarsa,1,2,50.00
Duk Operasi 80x90cm,Salah Pesan Customer,1,2,50.00
Glukometer Set,Salah Input Sistem,2,2,100.00
Gunting Bedah Bengkok 14cm,Salah Input Sistem,1,2,50.00


**Screening**  
Analysis at the product (SKU) level wasn't pursued further — the granularity is too fine. Out of 48 unique products, the maximum number of return cases for any single product is only 3, and most have just 1. The more meaningful pattern is already captured at the category level (Diagnostics and Laboratory show a concentration of SIS; Surgery & Procedures and Nursing & Inpatient Care show a concentration of SPC).

#### Cross-Tab / Confounding Check

(West Jakarta × customer type — checking whether these findings are actually related)

*Customer type breakdown, filtered to West Jakarta only.*

In [68]:
%%sql

SELECT
    customer_type_cleaned,
    COUNT (*) AS total_records
FROM
    return_cleaned
WHERE
    region_cleaned = 'Jakarta Barat'
GROUP BY
    customer_type_cleaned
ORDER BY
    total_records DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
3 rows affected.


customer_type_cleaned,total_records
RS Swasta,6
Klinik,4
RS Pemerintah,3


**Screening**  
n and row_pct match West Jakarta's numbers exactly (8/13) → verified in Section 2; result: independent (not a duplicate signal)

#### Category & Product

In [70]:
%%sql

WITH
    kategori_total AS (
        SELECT
            COUNT(*) AS total_loss
        FROM
            return_cleaned
    )

SELECT
    kategori,
    COUNT(*) AS total_loss_filter,
    kt.total_loss,
    ROUND((COUNT(*)::numeric / NULLIF(kt.total_loss,0)::numeric) * 100,2) AS row_pct
FROM
    return_cleaned rc
CROSS JOIN 
    kategori_total kt
GROUP BY
    kategori,
    kt.total_loss
ORDER BY
    total_loss_filter DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
5 rows affected.


kategori,total_loss_filter,total_loss,row_pct
Diagnostik,16,59,27.12
Consumable,13,59,22.03
Perawatan & Rawat Inap,11,59,18.64
Bedah & Tindakan,11,59,18.64
Laboratorium,8,59,13.56


**Screening**  
Range of 8-16 cases, with a shallow gap between categories (the largest difference across all 5 categories is just 8 points) → the distribution is fairly even, with no category dominating drastically.

*The product (SKU) level, top 10 by return count.*

In [71]:
%%sql

WITH
    product_total AS (
        SELECT
            COUNT(*) AS total_loss
        FROM
            return_cleaned
    )

SELECT
    sku_name,
    COUNT(*) AS total_loss_filter,
    pt.total_loss,
    ROUND((COUNT(*)::numeric / NULLIF(pt.total_loss,0)::numeric) * 100,2) AS ratio
FROM
    return_cleaned rc
CROSS JOIN 
    product_total pt
GROUP BY
    sku_name,
    pt.total_loss
ORDER BY
    total_loss_filter DESC
LIMIT 10

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
10 rows affected.


sku_name,total_loss_filter,total_loss,ratio
Syringe 1ml,3,59,5.08
Urine Stick 10 Parameter,3,59,5.08
Infus Set Anak,3,59,5.08
Selimut Pasien Rumah Sakit,3,59,5.08
Benang Catgut 2-0,3,59,5.08
IV Catheter 22G,2,59,3.39
Gunting Bedah Bengkok 14cm,2,59,3.39
NGT 16Fr,2,59,3.39
Tensimeter Manual Dewasa,2,59,3.39
Baju Pasien Dewasa,2,59,3.39


**Screening**  
The maximum count for any single product is only 3 (out of 48 unique SKUs), and most have just 1-2. The granularity is too fine to draw any conclusion about the distribution pattern — whether "dominant" or "even" — since one extra case could shift the ranking significantly. The product level isn't analyzed further; the category level is sufficient to answer BQ2.

**Verification:**  
**Trigger:** West Jakarta (region, SIS 8/13) and Government Hospital (customer type, SIS 8/13) show identical numbers → checked whether this is actually the same signal counted twice.

**Cross-tab:** West Jakarta × customer type → Private Hospital (6), Clinic (4), Government Hospital (3)

**Conclusion:** Government Hospital only accounts for 3 of the 13 cases in West Jakarta (not the majority) → these two findings are INDEPENDENT, not one overlapping signal. They can be reported as two separate findings.

#### Findings

System Input Error (Salah Input Sistem, SIS) is the most dominant reason code (47.46%), concentrated especially in West Jakarta (61.54%, +14.08 pts above baseline) and the Diagnostics/Laboratory categories (56-62%), with a sharp spike in Q2 (66.67%). The SIS rate relative to total transactions jumped 336% from Q1 to Q2 — far outpacing the transaction volume growth over the same period (14.8%) — which points to a cause beyond simple order volume; the specific root cause (a staffing or process change) can't be pinned down from the data available. Left unaddressed, this pattern will likely keep recurring above baseline, adding to the operational load of processing returns — especially for the team handling Diagnostics/Lab products in West Jakarta. Recommend investigating system change logs, input SOPs, or staffing around Q2, prioritizing the region and categories identified above.

Customer Ordering Error (Salah Pesan Customer, SPC) is the second-largest reason code (33.90%), with a different pattern from SIS — instead of a spike at one point in time, it's a trend that keeps worsening across the year: the SPC rate relative to total transactions rose consistently every quarter (1.34% → 1.75% → 2.23% → 5.39%), growing far faster than transaction volume over the same period (141% vs. 14% from Q3 to Q4). SPC is also concentrated in the Surgery & Procedures and Nursing & Inpatient Care categories (+11.55 pts above baseline each) and in the Clinic customer type (+8.21 pts) — likely tied to manual PO processes from customers that aren't automatically validated, though the specific cause behind this worsening trend can't be confirmed from the data available. At the regional level, SPC is also high in South Jakarta, but mixed in with an equally high SIS rate (50%/50%) — the relationship between the two hasn't been explored further. If left unaddressed, the return burden from SPC could keep growing faster than the business itself — a sign of a process problem, not just a natural consequence of more transactions. Recommend investigating the PO receiving/verification process, particularly for the Surgery & Procedures/Nursing & Inpatient Care categories and the Clinic customer type.

The distribution of returned product categories is fairly even (range of 8-16 cases), with no category dominating drastically; the product (SKU) level can't be analyzed further since the maximum count for any product is only 3 out of 48 unique SKUs. This even distribution actually reinforces the BQ1 finding — since returns aren't concentrated on specific products or categories, the root cause is unlikely to be product characteristics (e.g., a design or quality issue with a particular item), and more likely a process problem affecting all product lines equally — consistent with SIS (system input errors) and SPC (ordering mistakes from manual POs), both of which are process issues, not product issues. This means fixes targeted at specific products or categories (e.g., extra quality control for 1-2 SKUs) are unlikely to meaningfully reduce total return cases, since the problem is spread across every category. Recommend keeping the improvement focus on process (system input validation, PO verification) as recommended in BQ1, rather than product-specific interventions.

# 2. Financial Impact

*Total financial loss and average loss per return case, across the full dataset.*

In [64]:
%%sql

SELECT
    SUM(value_return) AS total_loss,
    AVG(value_return) AS avg
FROM
    return_cleaned


 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
1 rows affected.


total_loss,avg
171506500.0,2906889.8305084747


**Screening**  
Total financial loss from returns across 2024 reached IDR 171,506,500 from 59 return cases, averaging IDR 2,906,890 per case.

*Financial loss and case count broken down by category.*

In [72]:
%%sql

SELECT
    kategori,
    SUM(value_return) AS total_loss,
    COUNT(value_return) AS total_record
FROM
    return_cleaned
GROUP BY
    kategori
ORDER BY
    total_loss DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
5 rows affected.


kategori,total_loss,total_record
Diagnostik,66811000.0,16
Bedah & Tindakan,53336500.0,11
Perawatan & Rawat Inap,22706500.0,11
Consumable,21474000.0,13
Laboratorium,7178500.0,8


In [74]:
%%sql

WITH
    kategori_total AS (
        SELECT
            SUM(value_return) AS total_loss,
            COUNT(*) AS total_records
        FROM
            return_cleaned
    )

SELECT
    kategori,
    SUM(value_return) AS total_loss_filter,
    kt.total_loss,
    ROUND((SUM(value_return)::numeric / NULLIF(kt.total_loss,0)::numeric) * 100,2) AS loss_pct,
    ROUND(AVG(value_return):: numeric,2) AS average_loss_filter,
    COUNT(value_return) AS total_record_filter,
    kt.total_records,
    ROUND((COUNT(*)::numeric / NULLIF(total_records,0)::numeric) * 100,2) AS row_pct
FROM
    return_cleaned rc
CROSS JOIN 
    kategori_total kt
GROUP BY
    kategori,
    kt.total_loss,
    kt.total_records
ORDER BY
    total_loss_filter DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
5 rows affected.


kategori,total_loss_filter,total_loss,loss_pct,average_loss_filter,total_record_filter,total_records,row_pct
Diagnostik,66811000.0,171506500.0,38.96,4175687.50,16,59,27.12
Bedah & Tindakan,53336500.0,171506500.0,31.10,4848772.73,11,59,18.64
Perawatan & Rawat Inap,22706500.0,171506500.0,13.24,2064227.27,11,59,18.64
Consumable,21474000.0,171506500.0,12.52,1651846.15,13,59,22.03
Laboratorium,7178500.0,171506500.0,4.19,897312.50,8,59,13.56


**Screening**  
Diagnostics (n=16, 27.12% of cases) — 38.96% of total loss (IDR 66.8M), gap +11.84 pts vs. its case share, avg IDR 4,175,688/case → disproportionately costly, driven by volume (highest case count) → worth digging into

Surgery & Procedures (n=11, 18.64% of cases) — 31.10% of total loss (IDR 53.3M), gap +12.46 pts (the LARGEST gap of any category), avg IDR 4,848,773/case (the HIGHEST) → the MOST disproportionately costly, driven by value per case rather than volume → worth digging into

Nursing & Inpatient Care (n=11, 18.64% of cases) — 13.24% of total loss (IDR 22.7M), gap -5.40 pts, avg IDR 2,064,227/case → close to proportional → reasonably clean screening

Consumables (n=13, 22.03% of cases) — 12.52% of total loss (IDR 21.5M), gap -9.51 pts, avg IDR 1,651,846/case → disproportionately cheap (many cases, low value) → not a loss priority

Laboratory (n=8, 13.56% of cases) — 4.19% of total loss (IDR 7.2M), gap -9.37 pts, avg IDR 897,313/case (the LOWEST) → disproportionately cheap → not a loss priority

*Drilling into individual Diagnostics return lines — joining with invoice and return-quantity detail to identify which specific orders drive the loss.*

In [75]:
%%sql

SELECT
    rc.kategori,
    rc.sku_name,
    rc.value_return,
    id.unit_price,
    id.qty_delivered,
    rd.qty_returned,
    ROUND((rd.qty_returned::numeric / id.qty_delivered::numeric) * 100,2) AS row_pct
FROM
    return_cleaned rc
LEFT JOIN
    invoice_detail id
    ON rc.invoice_id = id.invoice_id AND rc.sku_name = id.nama_produk
LEFT JOIN
    return_data rd
    ON rc.retur_id = rd.retur_id
WHERE
    rc.kategori = 'Diagnostik'
ORDER BY
    rc.value_return DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
16 rows affected.


kategori,sku_name,value_return,unit_price,qty_delivered,qty_returned,row_pct
Diagnostik,Urine Stick 10 Parameter,18473000.0,318500.0,191,58,30.37
Diagnostik,Rapid Test HbsAg,15979000.0,420500.0,173,38,21.97
Diagnostik,Rapid Test HbsAg,8195000.0,372500.0,73,22,30.14
Diagnostik,Glukometer Set,6300000.0,252000.0,113,25,22.12
Diagnostik,Pulse Oximeter,4650000.0,310000.0,65,15,23.08
Diagnostik,Stethoscope Anak,2761000.0,251000.0,165,11,6.67
Diagnostik,Tensimeter Manual Dewasa,1908000.0,238500.0,42,8,19.05
Diagnostik,Stethoscope Dewasa,1834000.0,262000.0,58,7,12.07
Diagnostik,Glukometer Set,1680000.0,280000.0,47,6,12.77
Diagnostik,Timbangan Bayi Digital,1587000.0,529000.0,16,3,18.75


*Drilling down, for Surgery & Procedures.*

In [76]:
%%sql

SELECT
    rc.kategori,
    rc.sku_name,
    rc.value_return,
    id.unit_price,
    id.qty_delivered,
    rd.qty_returned,
    ROUND((rd.qty_returned::numeric / id.qty_delivered::numeric) * 100,2) AS row_pct
FROM
    return_cleaned rc
LEFT JOIN
    invoice_detail id
    ON rc.invoice_id = id.invoice_id AND rc.sku_name = id.nama_produk
LEFT JOIN
    return_data rd
    ON rc.retur_id = rd.retur_id
WHERE
    rc.kategori = 'Bedah & Tindakan'
ORDER BY
    rc.value_return DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
11 rows affected.


kategori,sku_name,value_return,unit_price,qty_delivered,qty_returned,row_pct
Bedah & Tindakan,Handscoon Steril 6.5,15645000.0,447000.0,168,35,20.83
Bedah & Tindakan,Handscoon Steril 7.5,12801000.0,376500.0,118,34,28.81
Bedah & Tindakan,Benang Catgut 2-0,10678500.0,395500.0,81,27,33.33
Bedah & Tindakan,Benang Catgut 2-0,5537000.0,395500.0,70,14,20.00
Bedah & Tindakan,Gunting Bedah Bengkok 14cm,1852500.0,123500.0,79,15,18.99
Bedah & Tindakan,Benang Catgut 2-0,1760500.0,251500.0,113,7,6.19
Bedah & Tindakan,Klem Arteri Bengkok,1440000.0,160000.0,43,9,20.93
Bedah & Tindakan,Duk Operasi 80x90cm,1225500.0,64500.0,63,19,30.16
Bedah & Tindakan,Duk Operasi 80x90cm,1067000.0,48500.0,94,22,23.40
Bedah & Tindakan,Pinset Bedah 14cm,912000.0,57000.0,74,16,21.62


*Checking the reason code for the two Sterile Gloves SKUs flagged as outliers above.*

In [77]:
%%sql

SELECT
    return_reason_cleaned,
    COUNT(*) AS return_count
FROM
    return_cleaned
WHERE
    sku_name = 'Handscoon Steril 6.5' OR
    sku_name = 'Handscoon Steril 7.5'
GROUP BY
    return_reason_cleaned

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
2 rows affected.


return_reason_cleaned,return_count
Salah Input Sistem,1
Salah Pesan Customer,1


*Reason code breakdown restricted to the two categories driving the largest financial loss (Diagnostics, Surgery & Procedures).*

In [78]:
%%sql

SELECT
    kategori,
    return_reason_cleaned,
    COUNT(*) AS return_count
FROM
    return_cleaned
WHERE
    kategori = 'Bedah & Tindakan' OR
    kategori = 'Diagnostik'
GROUP BY
    return_reason_cleaned,
    kategori

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
7 rows affected.


kategori,return_reason_cleaned,return_count
Diagnostik,Salah Pesan Customer,4
Diagnostik,Kadaluarsa,1
Diagnostik,Salah Input Sistem,9
Bedah & Tindakan,Kadaluarsa,1
Bedah & Tindakan,Salah Input Sistem,5
Bedah & Tindakan,Salah Pesan Customer,5
Diagnostik,Barang Rusak,2


*The 5 largest individual return cases by value, across the entire dataset — used to check how concentrated the total loss really is.*

In [79]:
%%sql

SELECT 
    sku_name, 
    kategori,
    return_reason_cleaned,
    value_return
FROM 
    return_cleaned
ORDER BY 
    value_return DESC
LIMIT 5

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
5 rows affected.


sku_name,kategori,return_reason_cleaned,value_return
Urine Stick 10 Parameter,Diagnostik,Salah Pesan Customer,18473000.0
Rapid Test HbsAg,Diagnostik,Barang Rusak,15979000.0
Handscoon Steril 6.5,Bedah & Tindakan,Salah Input Sistem,15645000.0
IV Catheter 22G,Consumable,Salah Pesan Customer,14563500.0
Handscoon Steril 7.5,Bedah & Tindakan,Salah Pesan Customer,12801000.0


### **Findings**

Total financial loss from returns across 2024 reached **IDR 171,506,500** from 59 return cases — averaging IDR 2,906,890 per case. But the distribution is highly skewed (see Finding #3): just 5 cases (8.5%) account for 45% of this total.

Diagnostics accounts for the largest financial loss (IDR 66,811,000, 38.96% of the IDR 171,506,500 total), disproportionate to its 27.12% share of total cases — a +11.84 pt gap. This loss is driven by a combination of the highest case count (16 of 59) and 2 high-volume outliers: Urine Stick 10 Parameter (IDR 18,473,000, 58 units, reason: SPC) and Rapid Test HbsAg (IDR 15,979,000, 38 units, reason: Damaged Goods) — together accounting for 63.83% of this category's total loss. One of these outliers (Urine Stick) ties directly back to the SPC pattern identified in BQ1. Since the loss is concentrated in a handful of high-volume transactions rather than spread evenly across the category, an intervention targeting the whole Diagnostics category (e.g., blanket quality control) is likely less efficient than a control aimed specifically at high-volume orders. Recommend adding verification/approval for large-quantity orders (particularly tied to the input/ordering process, in line with the BQ1 recommendation), plus a separate review of the Rapid Test HbsAg case since it's a damaged-goods issue rather than a process error.

Surgery & Procedures accounts for IDR 53,336,500 in losses (31.10% of the total), disproportionate to its 18.64% case share — a +12.46 pt gap, the largest of any category. The average loss per case (IDR 4,848,773) is misleading: 3 of the 11 cases (Sterile Gloves 6.5, IDR 15,645,000/35 units/SIS; Sterile Gloves 7.5, IDR 12,801,000/34 units/SPC; Catgut Suture 2-0, IDR 10,678,500/27 units) account for 73.35% of this category's loss (top-3 average IDR 13,041,500 vs. just IDR 1,776,500 for the remaining 8). 2 of these 3 outliers are already confirmed as process-related (SIS/SPC), consistent with the BQ1 finding that this category shows an SPC concentration +11.55 pts above baseline. Since the loss is concentrated in a small number of high-volume transactions, a blanket quality-control measure for this category is likely less efficient than a control based on transaction quantity. Recommend verification/approval for large-quantity orders, in line with the BQ1 recommendation and Finding #1.

Across all categories, 5 of the 59 return cases (8.5%) account for IDR 77,461,500 (45.15%) of the total loss — spread across 3 different categories (Diagnostics ×2, Surgery & Procedures ×2, Consumables ×1: IV Catheter 22G, IDR 14,563,500). 4 of these 5 cases are process-related (1 SIS, 3 SPC), with only 1 tied to Damaged Goods — and the 4 cases with verified quantities (Urine Stick 58 units, Rapid Test 38 units, Sterile Gloves 6.5 35 units, Sterile Gloves 7.5 34 units) are all high-volume. This shows the root of the loss isn't a specific category or product, but high-volume transactions slipping through without validation — including in the Consumables category, which earlier screening had flagged as "not a priority" (a blind-spot risk if mitigation only focuses on Diagnostics/Surgery & Procedures). Recommend adding a quantity-threshold control/approval per transaction (rather than per category), to complement the process recommendations in BQ1 and Findings #1-2.

# 3. Who & Where

*Return count and % share by customer type.*

In [80]:
%%sql
WITH total AS
(
    SELECT
        COUNT(*) AS total_records
    FROM
        return_cleaned
)

SELECT
    rc.customer_type_cleaned,
    COUNT(*) AS total_record,
    tl.total_records AS total_data,
    ROUND((COUNT(*)::numeric / NULLIF(tl.total_records,0)::numeric) * 100,2) AS rate
FROM
    return_cleaned rc
CROSS JOIN
    total tl
GROUP BY
    rc.customer_type_cleaned,
    total_data
ORDER BY
    total_record DESC


 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
3 rows affected.


customer_type_cleaned,total_record,total_data,rate
RS Swasta,27,59,45.76
Klinik,19,59,32.20
RS Pemerintah,13,59,22.03


**Screening**  
Private Hospital (45.76%) is the customer type with the most returns — clearly ahead of Clinic (32.20%) and Government Hospital (22.03%). The ranking has a clear gap between tiers, not two close competitors.

*Reason code composition within each customer type.*

In [82]:
%%sql
WITH customer_reason AS (
    SELECT
        customer_type_cleaned,
        return_reason_cleaned,
        COUNT(*) AS return_count
    FROM
        return_cleaned  
    GROUP BY
        customer_type_cleaned,
        return_reason_cleaned
)

SELECT
    customer_type_cleaned,
    return_reason_cleaned,
    return_count,
    SUM(return_count) OVER (PARTITION BY customer_type_cleaned) AS total_per_customer_type,
        ROUND(
            return_count * 100.0 / SUM(return_count) OVER (PARTITION BY customer_type_cleaned), 
            2
        ) AS row_pct
FROM
    customer_reason

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
12 rows affected.


customer_type_cleaned,return_reason_cleaned,return_count,total_per_customer_type,row_pct
Klinik,Barang Rusak,1,19,5.26
Klinik,Kadaluarsa,1,19,5.26
Klinik,Salah Input Sistem,9,19,47.37
Klinik,Salah Pesan Customer,8,19,42.11
RS Pemerintah,Barang Rusak,2,13,15.38
RS Pemerintah,Kadaluarsa,1,13,7.69
RS Pemerintah,Salah Input Sistem,8,13,61.54
RS Pemerintah,Salah Pesan Customer,2,13,15.38
RS Swasta,Barang Rusak,4,27,14.81
RS Swasta,Kadaluarsa,2,27,7.41


*Return count and % share by region.*

In [83]:
%%sql
WITH total AS
(
    SELECT
        COUNT(*) AS total_records
    FROM
        return_cleaned
)

SELECT
    rc.region_cleaned,
    COUNT(*) AS total_record,
    tl.total_records AS total_data,
    ROUND((COUNT(*)::numeric / NULLIF(tl.total_records,0)::numeric) * 100,2) AS rate
FROM
    return_cleaned rc
CROSS JOIN
    total tl
GROUP BY
    rc.region_cleaned,
    total_data
ORDER BY
    total_record DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
9 rows affected.


region_cleaned,total_record,total_data,rate
Jakarta Barat,13,59,22.03
Bekasi,10,59,16.95
Jakarta Selatan,10,59,16.95
Jakarta Timur,6,59,10.17
Jakarta Utara,5,59,8.47
Tangerang,5,59,8.47
Bogor,4,59,6.78
Depok,3,59,5.08
Jakarta Pusat,3,59,5.08


*Reason code composition within each region.*

In [84]:
%%sql

WITH category_count AS (
    SELECT
        region_cleaned,
        return_reason_cleaned,
        COUNT (*) AS total_records
    FROM
        return_cleaned
    GROUP BY
        region_cleaned,
        return_reason_cleaned
)

SELECT
    region_cleaned,
    return_reason_cleaned,
    total_records,
    SUM(total_records) OVER (PARTITION BY region_cleaned) AS total_per_category,
    ROUND (total_records * 100.0 / SUM(total_records) OVER (PARTITION BY region_cleaned),2) AS row_pct
FROM
    category_count
ORDER BY
    region_cleaned,
    total_records DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
25 rows affected.


region_cleaned,return_reason_cleaned,total_records,total_per_category,row_pct
Bekasi,Salah Input Sistem,5,10,50.00
Bekasi,Salah Pesan Customer,4,10,40.00
Bekasi,Kadaluarsa,1,10,10.00
Bogor,Salah Input Sistem,2,4,50.00
Bogor,Salah Pesan Customer,2,4,50.00
Depok,Kadaluarsa,1,3,33.33
Depok,Barang Rusak,1,3,33.33
Depok,Salah Input Sistem,1,3,33.33
Jakarta Barat,Salah Input Sistem,8,13,61.54
Jakarta Barat,Salah Pesan Customer,3,13,23.08


#### Findings

Private Hospital is the customer type with the most returns (45.76%, 27 of 59 cases) — clearly ahead of Clinic (32.20%) and Government Hospital (22.03%), with a clear tiered ranking rather than close competitors. However, the reason code composition within Private Hospital doesn't deviate meaningfully from the BQ1 baseline (SIS 40.74% vs. baseline 47.46%, -6.72 pts; SPC 37.04% vs. baseline 33.90%, +3.14 pts) — verified using the existing BQ1 screening data, without any new query. This suggests Private Hospital's dominance isn't caused by a specific process problem (SIS/SPC), but is most likely simply because Private Hospital is the largest segment by transaction volume overall — total transaction data per customer type wasn't available to confirm this further. Since there's no process signal that needs specific mitigation in this segment, there's no additional recommendation for Private Hospital beyond what's already covered in BQ1 (general process improvement, not customer-type-specific).

Nine regions form a top cluster that's clearly ahead of the rest: West Jakarta (22.03%), Bekasi (16.95%), and South Jakarta (16.95%) — together 55.93% versus 44.07% for the remaining 6 regions, with the biggest gap (6.78 pts) falling between South Jakarta and East Jakarta. This signal is moderate — not as strong as the SIS+SPC cluster in BQ1 (81%/19%) — so it's noted as a tendency rather than a decisive dominance. The three top-cluster regions already have their own status from BQ1 (not re-explored here): West Jakarta shows a specific SIS concentration (61.54%, verified as independent from the Government Hospital finding); Bekasi is a clean screening; South Jakarta has both SIS and SPC equally high (50%/50%), with the cause still unidentified (parked for later). The remaining six regions are flat — either a clean screening or too small an n to draw conclusions. Since the root causes for the priority regions are already covered by the BQ1 recommendations (improving the input process/PO verification, particularly in West Jakarta), there's no additional recommendation beyond that — except closing out the still-open South Jakarta investigation.

# 4. Trend Over Time

*Total returns per quarter.*

In [85]:
%%sql

SELECT
    CASE
        WHEN EXTRACT(MONTH FROM return_date) BETWEEN 1 AND 3 THEN 'Q1'
        WHEN EXTRACT(MONTH FROM return_date) BETWEEN 4 AND 6 THEN 'Q2'
        WHEN EXTRACT(MONTH FROM return_date) BETWEEN 7 AND 9 THEN 'Q3'
        WHEN EXTRACT(MONTH FROM return_date) BETWEEN 10 AND 12 THEN 'Q4'
    END AS quarter,
    COUNT(*) AS total_data
FROM 
    return_cleaned
GROUP BY 
    quarter
ORDER BY 
    quarter

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
4 rows affected.


quarter,total_data
Q1,8
Q2,15
Q3,14
Q4,22


*Total invoices per quarter — the denominator for the return-rate calculation.*

In [86]:
%%sql

SELECT
    CASE
        WHEN EXTRACT(MONTH FROM invoice_date) BETWEEN 1 AND 3 THEN 'Q1'
        WHEN EXTRACT(MONTH FROM invoice_date) BETWEEN 4 AND 6 THEN 'Q2'
        WHEN EXTRACT(MONTH FROM invoice_date) BETWEEN 7 AND 9 THEN 'Q3'
        WHEN EXTRACT(MONTH FROM invoice_date) BETWEEN 10 AND 12 THEN 'Q4'
    END AS quarter,
    COUNT(*) AS total_data
FROM 
    invoice_data
GROUP BY 
    quarter
ORDER BY 
    quarter

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
4 rows affected.


quarter,total_data
Q1,149
Q2,171
Q3,179
Q4,204
